adapta/storage/blob/s3_storage_client.py

```diff
@@ -132,15 +148,53 @@ def download_blobs(
         filter_predicate: Optional[Callable[[...], bool]] = None,
     ) -> None:
         """
-        Not yet implemented in S3 Client
+        Downloads blobs from S3 storage to a local path.
+
+        :param blob_path: Path to blob
+        :param local_path: Local path to download the blobs to
+        :param threads: Number of threads to use for the download
+        :param filter_predicate: Optional callable to filter blobs
+        :return:
         """
-        raise NotImplementedError("Not yet implemented in S3StorageClient")
+        s3_path = cast_path(blob_path)
+        try:
+            blobs = self._s3_resource.Bucket(s3_path.bucket).objects.filter(Prefix=s3_path.path)
+            for blob in blobs:
+                if filter_predicate is None or filter_predicate(blob):
+                    local_file_path = f"{local_path}/{blob.key}"
+                    os.makedirs(os.path.dirname(local_file_path), exist_ok=True)
+                    self._s3_resource.meta.client.download_file(s3_path.bucket, blob.key, f"{local_path}/{blob.key}")
+        except StorageClientError as error:
+            print(f"Error downloading blob: {error}")
 
     def copy_blob(self, blob_path: DataPath, target_blob_path: DataPath, doze_period_ms: int) -> None:
         """
-        Not implemented in S3 Client
+        Copies a blob from one location to another in S3 storage.
+
+        :param blob_path: Path to the source blob
+        :param target_blob_path: Path to the target location
+        :param doze_period_ms: Time to sleep between operations in milliseconds
         """
-        raise NotImplementedError("Not implemented in S3StorageClient")
+        source_s3_path = cast_path(blob_path)
+        target_s3_path = cast_path(target_blob_path)
+
+        source_objects = self._s3_resource.Bucket(source_s3_path.bucket).objects.filter(Prefix=source_s3_path.path)
+
+        for source_object in source_objects:
+            # If the source path is a directory, construct the target object key
+            # If the source path is a file, the target object key is the target path
+            if source_s3_path.path == source_object.key:
+                target_object_key = target_s3_path.path
+            else:
+                target_object_key = source_object.key.replace(source_s3_path.path, target_s3_path.path, 1)
+
+            copy_source = {"Bucket": source_s3_path.bucket, "Key": source_object.key}
+
+            try:
+                self._s3_resource.meta.client.copy(copy_source, target_s3_path.bucket, target_object_key)
+                time.sleep(doze_period_ms / 1000.0)
```

Comment:
if this is s3>s3 copy, you need a loop to wait for the copy to complete, otherwise this sleep is kind of pointless